### DISCUSSION TITLE
Plugin Guidelines

### DISCUSSION THREAD

**Discussion by rikgirbes** [saadeghi/daisyui#1014] [Ideas / feature requests]
In the pursuit of making daisyUI better, there were multiple discussions regarding more functionality. (#1004)
Most of us, including @saadeghi, believed daisyUI should remain a `design`-only framework.
Hence: plugins

They may contain Javascript / other frameworks / etc. 
When a plugin is approved, it might be added to a new section of the daisyUI docs, under (probably) 'plugins'.

In this discussion we'd like to think about/setup, basic guidelines for plugins.

Some previously mentioned rules/guidelines:

- Vanilla-js/ts only
- Only (mostly?) using Tailwindcss/daisyUI classes (through `@apply` or directly in markup)
- Shouldn't interfering with other basic styling
- **Optional** framework wrappers (e.g. React components / svelte / Vue etc)

When we've collected more thoughts, I (or someone else) will setup a basic guide and put up a PR 😊 
Thanks!

**Comment by Darkhorse-Fraternity** [comment]
Recommended:
1. Separate Logic and View  eg:  [table-core](https://github.com/TanStack/table/tree/main/packages/table-core)  
2. View Binding: https://github.com/themesberg/flowbite/blob/main/src/plugins/datepicker.js
3. ts

**Comment by ThatXliner** [comment]
1. What about web components?
2. Or Vanilla JS? They should probably take in an element and stuff with it on runtime.

**Comment by saadeghi** [comment]
I think having a strict guideline for **_how plugins should be written_** (language, stack, methods, design, etc...) would not work 🤔  
Instead I'm looking to **_the end results_**.

Because plugins should be completely separate projects, owned by their creators. So they can be anything...  
- It can be a NPM package that adds 1 custom theme to daisyUI (written with any tools or languages or methods)
- It can be a file that applies a specific style to buttons (for example Ripple effect 🤷‍♂️ )
- It can modify some components by overriding CSS for people who want their own design decisions
- It can be a JS file that adds/removes style for specific elements (like a datepicker)
There's no limit.

As long as the plugin's end result looks good I will list them on the site and I will recommend them to daisyUI users.  
So if I have to put some rules, they would be:  
- Plugin must be easy to install and use
- Plugin should be well documented (on its GitHub repo or custom web page) so people would know what it does and how they should use it.
- Plugin must do a specific job or add a specific value, that a group of people can benefit from it.

**Reply by rikgirbes** [reply]
Ok, in that case, I'd think the docs on daisyUI should be well categorised.
E.g. react plugins / headless / themes / etc.

Otherwise it'll be a list of all things just mashed together and scrambled.


But then again, why not create some default "template" for plugins, and its docs, that could essentially also be added to the daisyUI docs itself.?

I'm just `scared` of how normally these things turn out, if you leave it all to everyones imagination. (read: react plugins everywhere, jquery plugins everywhere, etc)
Thats why I thought some guidelines might be a good idea, maybe even a default docs template.

**Reply by saadeghi** [reply]
I understand.  
I will of-course categorize them when the number of plugins goes high.  
But do not worry about having a default template or a guideline. This won't be a problem with open-source projects because as soon as people start using a package, there will be issues, questions, discussions and contributions that will lead the repo to improvement (both with the code and the docs)  

After all, a daisyUI plugin could be as small as a theme.  
**For example:**  
This is `mytheme` theme for daisyUI. how to use? Copy below code to your `tailwind.config.js`. The end.
```js
daisyui: {
  themes: [
    {
      mytheme: {
        "primary": "#ffc9db",
        "secondary": "#b0e0f2",
        "accent": "#f7b3b2",
        "neutral": "#181C2A",
        "base-100": "#E1E1EA",
        "info": "#41CBF1",
        "success": "#155B4E",
        "warning": "#F98C10",
        "error": "#F14037",
      },
    },
  ],
},
```
That's all that example plugin needs. one `readme.md` file would be enough for it.

And it's not really possible to have a temple. As I said, plugins could be anything. Could be made for any specific JS library with any stack.

However If you need help with documenting your upcoming datepicker plugin or its file structure, etc, let me know and I will help as much as I can.
